### PR TITLE
Reuse scratch memory in SYCL TeamPolicy

### PR DESCRIPTION
**Pull request by masterleinad**
Similar to #3873, we can reuse the scratch memory allocated for `TeamPolicy`.

### CODE REVIEW

File: core/src/SYCL/Kokkos_SYCL_Instance.cpp [hunk 1]
```diff
@@ -162,6 +163,27 @@ void SYCLInternal::initialize(const sycl::queue& q) {
...
+            "SYCLDeviceUSMSpace::ScratchMemory", m_team_scratch_current_size);
+  }
+  if ((bytes > m_team_scratch_current_size) ||
+      ((bytes < m_team_scratch_current_size) && (force_shrink))) {
+    m_team_scratch_current_size = bytes;
+    m_team_scratch_ptr =
+        Kokkos::kokkos_realloc<Experimental::SYCLDeviceUSMSpace>(
+            m_team_scratch_ptr, m_team_scratch_current_size);
+  }
+  return m_team_scratch_ptr;
```
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Instance.cpp:186]: This pretty much verbatim copied from the `Cuda` implementation.

**Comment by masterleinad** [comment]
The whole implementation is pretty much taken from the `Cuda` backend.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp [hunk 2]
```diff
@@ -788,11 +779,15 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
...
-        sizeof(char) * m_scratch_size[1] * m_league_size, q);
+    const auto& space = *m_policy.space().impl_internal_space_instance();
+    m_scratch_ptr[0]  = nullptr;
+    m_scratch_ptr[1]  = m_team_size <= 0
+                           ? nullptr
+                           : m_policy.space()
+                                 .impl_internal_space_instance()
+                                 ->resize_team_scratch_space(
+                                     static_cast<ptrdiff_t>(m_scratch_size[1]) *
+                                     m_league_size);
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:790]: These are `int` right?  Why do you cast?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp [hunk 3]
```diff
@@ -788,11 +779,15 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
...
     // Functor's reduce memory, team scan memory, and team shared memory depend
     // upon team size.
-    const auto& space    = *m_policy.space().impl_internal_space_instance();
-    const sycl::queue& q = *space.m_queue;
-    m_scratch_ptr[0]     = nullptr;
-    m_scratch_ptr[1]     = sycl::malloc_device(
-        sizeof(char) * m_scratch_size[1] * m_league_size, q);
+    const auto& space = *m_policy.space().impl_internal_space_instance();
+    m_scratch_ptr[0]  = nullptr;
+    m_scratch_ptr[1]  = m_team_size <= 0
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:784]: I don't get the ternary operator here.  Did you mean to check if `m_league_size` is non negative?

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:784]: It doesn't matter here (but it did for the `Cuda` backend). Let me just remove it.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:790]: Purely conformity with the `Cuda` implementation. I can change that if you want.

File: core/src/SYCL/Kokkos_SYCL_Instance.cpp [hunk 4]
```diff
@@ -162,6 +163,27 @@ void SYCLInternal::initialize(const sycl::queue& q) {
...
+  m_team_scratch_ptr          = nullptr;
+}
+
+void* SYCLInternal::resize_team_scratch_space(std::int64_t bytes,
+                                              bool force_shrink) {
+  if (m_team_scratch_current_size == 0) {
+    m_team_scratch_current_size = bytes;
+    m_team_scratch_ptr =
+        Kokkos::kokkos_malloc<Experimental::SYCLDeviceUSMSpace>(
+            "SYCLDeviceUSMSpace::ScratchMemory", m_team_scratch_current_size);
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.cpp]: 272dd5597301e8568db473ddc8651b962d56c781

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp (hunk 3 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp]: 272dd5597301e8568db473ddc8651b962d56c781

**Review by nliber — APPROVED** [review]

**Comment by dalg24** [comment]
Plz rebase

**Comment by masterleinad** [comment]
> Plz rebase

Done.

File: core/src/SYCL/Kokkos_SYCL_Instance.cpp [hunk 5]
```diff
@@ -42,6 +42,7 @@
 //@HEADER
 */
 
+#include <Kokkos_Core.hpp>
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Instance.cpp:45]: Why are you including `<Kokkos_Core.hpp>`?

**Review by dalg24 — COMMENTED** [review]
Looks good other than the header includes

File: core/src/SYCL/Kokkos_SYCL_Instance.cpp (hunk 5 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.cpp:45]: This is for `kokkos_malloc`. I could try forward declaring in `Kokkos_core_fwd.hpp` instead.

Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL_Instance.cpp:45]: It just makes little sense to include everything and below include selected headers.

**Pull request merged**